clients/src/main/java/org/apache/kafka/common/internals/SecurityManagerCompatibility.java

```diff
@@ -0,0 +1,355 @@
+                if (cause instanceof RuntimeException) {
+                    throw (RuntimeException) cause;
+                } else {
+                    throw new RuntimeException(cause);
+                }
+            }
+        }
+
+        /**
+         * @param context The current AccessControlContext
+         * @return The result of Subject.getSubject(AccessControlContext)
+         */
+        private Subject getSubject(Object context) {
+            try {
+                return (Subject) getSubject.invoke(null, context);
+            } catch (IllegalAccessException e) {
+                throw new UnsupportedOperationException(e);
+            } catch (InvocationTargetException e) {
+                Throwable cause = e.getCause();
+                if (cause instanceof RuntimeException) {
+                    throw (RuntimeException) cause;
+                } else {
+                    throw new RuntimeException(cause);
+                }
+            }
+        }
+
+        @Override
+        public Subject current() {
+            return getSubject(getContext());
+        }
+
+        /**
+         * @return The result of Subject.doAs(Subject, PrivilegedExceptionAction)
+         */
+        private <T> T doAs(Subject subject, PrivilegedExceptionAction<T> action) throws PrivilegedActionException {
+            try {
+                return (T) doAs.invoke(null, subject, action);
+            } catch (IllegalAccessException e) {
+                throw new UnsupportedOperationException(e);
+            } catch (InvocationTargetException e) {
+                Throwable cause = e.getCause();
+                if (cause instanceof PrivilegedActionException) {
+                    throw (PrivilegedActionException) cause;
+                } else if (cause instanceof RuntimeException) {
+                    throw (RuntimeException) cause;
+                } else  {
+                    throw new RuntimeException(cause);
+                }
+            }
+        }
+
+        @Override
+        public <T> T callAs(Subject subject, Callable<T> callable) throws CompletionException {
+            try {
+                return doAs(subject, callable::call);
+            } catch (PrivilegedActionException e) {
+                throw new CompletionException(e.getCause());
+            }
+        }
+    }
+
+    /**
+     * This class implements reflective access to the methods of Subject added to replace deprecated methods.
+     * <p>Instantiating this class may fail if any of the required classes or methods are not found.
+     * Method invocations for this class may fail with {@link UnsupportedOperationException} if all methods are found,
+     * but the operation is not permitted to be invoked.
+     * <p>This class is expected to be instantiable in JRE >= 18. At the time of writing, these methods do not have
+     * a sunset date, and are expected to be available past the removal of the SecurityManager.
+     */
```

Comment:
Might be nice to add a `@see` tag that points at the new APIs that aren't available in older JDKs, and/or the JEP(s) that deprecate older APIs or introduce the modern ones.

Reply:
I originally left out all links to the underlying methods because I thought they would prevent compilation. TIL that javadocs in non-public-API classes can be "broken" with bad links because they're never processed by :javadoc.